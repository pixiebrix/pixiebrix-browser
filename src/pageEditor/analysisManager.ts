/*
 * Copyright (C) 2023 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import BlockTypeAnalysis from "@/analysis/analysisVisitors/blockTypeAnalysis";
import ExtensionUrlPatternAnalysis from "@/analysis/analysisVisitors/extensionUrlPatternAnalysis";
import OutputKeyAnalysis from "@/analysis/analysisVisitors/outputKeyAnalysis";
import RenderersAnalysis from "@/analysis/analysisVisitors/renderersAnalysis";
import TemplateAnalysis from "@/analysis/analysisVisitors/templateAnalysis";
import TraceAnalysis from "@/analysis/analysisVisitors/traceAnalysis";
import ReduxAnalysisManager from "@/analysis/ReduxAnalysisManager";
import { type UUID } from "@/core";
import { type TraceRecord } from "@/telemetry/trace";
import { isAnyOf, type PayloadAction } from "@reduxjs/toolkit";
import { type RootState } from "./pageEditorTypes";
import { actions as editorActions } from "@/pageEditor/slices/editorSlice";
import runtimeSlice from "./slices/runtimeSlice";
import RequestPermissionAnalysis from "@/analysis/analysisVisitors/requestPermissionAnalysis";
import FormBrickAnalysis from "@/analysis/analysisVisitors/formBrickAnalysis";
import { selectActiveElementTraces } from "./slices/runtimeSelectors";
import VarAnalysis from "@/analysis/analysisVisitors/varAnalysis/varAnalysis";
import analysisSlice from "@/analysis/analysisSlice";
import RegexAnalysis from "@/analysis/analysisVisitors/regexAnalysis";

const runtimeActions = runtimeSlice.actions;

const pageEditorAnalysisManager = new ReduxAnalysisManager();

// These actions will be used with every analysis so the annotation path is up-to-date
// with the node position in the pipeline
const nodeListMutationActions = [
  editorActions.addNode,
  editorActions.moveNode,
  editorActions.removeNode,
] as const;

// The order in which the analysis are registered is important.
// The first analysis registered will be the first to run.
// When multiple actions (e.g. typing) trigger analysis, the later analysis have more chances to get aborted.
// Try to put the faster analysis first, and the slower ones at the end.

pageEditorAnalysisManager.registerAnalysisEffect(
  (
    action: PayloadAction<{ extensionId: UUID; records: TraceRecord[] }>,
    state: RootState
  ) => {
    // TraceAnalysis filter the trace errors, thus
    // selecting all records here to avoid double filtering
    const records = selectActiveElementTraces(state);

    return new TraceAnalysis(records);
  },
  {
    matcher: isAnyOf(
      runtimeActions.setExtensionTrace,
      ...nodeListMutationActions
    ),
  }
);

pageEditorAnalysisManager.registerAnalysisEffect(
  () => new BlockTypeAnalysis(),
  {
    matcher: isAnyOf(...nodeListMutationActions),
  }
);

pageEditorAnalysisManager.registerAnalysisEffect(
  () => new FormBrickAnalysis(),
  {
    matcher: isAnyOf(...nodeListMutationActions),
  }
);

pageEditorAnalysisManager.registerAnalysisEffect(
  () => new RenderersAnalysis(),
  {
    matcher: isAnyOf(...nodeListMutationActions),
  }
);

pageEditorAnalysisManager.registerAnalysisEffect(() => new TemplateAnalysis(), {
  matcher: isAnyOf(editorActions.editElement, ...nodeListMutationActions),
});

pageEditorAnalysisManager.registerAnalysisEffect(
  () => new ExtensionUrlPatternAnalysis(),
  {
    matcher: isAnyOf(editorActions.editElement, ...nodeListMutationActions),
  }
);

pageEditorAnalysisManager.registerAnalysisEffect(
  () => new RequestPermissionAnalysis(),
  {
    matcher: isAnyOf(editorActions.editElement, ...nodeListMutationActions),
  }
);

pageEditorAnalysisManager.registerAnalysisEffect(() => new RegexAnalysis(), {
  matcher: isAnyOf(editorActions.editElement, ...nodeListMutationActions),
});

const varAnalysisFactory = (
  action: PayloadAction<{ extensionId: UUID; records: TraceRecord[] }>,
  state: RootState
) => {
  const records = selectActiveElementTraces(state);
  return new VarAnalysis(records);
};

// OutputKeyAnalysis seems to be the slowest one, so we register it in the end
pageEditorAnalysisManager.registerAnalysisEffect(
  () => new OutputKeyAnalysis(),
  {
    matcher: isAnyOf(editorActions.editElement, ...nodeListMutationActions),
  }
);

// VarAnalysis is not the slowest itself, but it triggers a post-analysis action,
// so it is the last one
pageEditorAnalysisManager.registerAnalysisEffect(
  varAnalysisFactory,
  {
    matcher: isAnyOf(
      editorActions.editElement,
      runtimeActions.setExtensionTrace,
      ...nodeListMutationActions
    ),
  },
  {
    postAnalysisAction(analysis, extensionId, listenerApi) {
      listenerApi.dispatch(
        analysisSlice.actions.setKnownVars({
          extensionId,
          vars: analysis.getKnownVars(),
        })
      );
    },
  }
);

export default pageEditorAnalysisManager;
