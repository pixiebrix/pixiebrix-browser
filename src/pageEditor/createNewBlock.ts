/*
 * Copyright (C) 2022 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import { BlockConfig } from "@/blocks/types";
import { defaultBlockConfig } from "@/blocks/util";
import { RegistryId, Schema } from "@/core";
import { uuidv4 } from "@/types/helpers";
import { getExampleBlockConfig } from "./exampleBlockConfigs";

export function createNewBlock(
  blockId: RegistryId,
  blockInputSchema?: Schema
): BlockConfig {
  return {
    id: blockId,
    instanceId: uuidv4(),
    config:
      getExampleBlockConfig(blockId) ??
      (blockInputSchema == null ? {} : defaultBlockConfig(blockInputSchema)),
  };
}
