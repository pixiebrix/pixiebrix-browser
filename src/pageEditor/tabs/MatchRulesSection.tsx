/*
 * Copyright (C) 2023 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import UrlPatternField from "@/pageEditor/fields/UrlPatternField";
import { makeLockableFieldProps } from "@/pageEditor/fields/makeLockableFieldProps";
import SelectorMatchField from "@/pageEditor/fields/SelectorMatchField";
import FieldSection from "@/pageEditor/fields/FieldSection";
import React from "react";

const MatchRulesSection: React.FunctionComponent<{
  isLocked: boolean;
}> = ({ isLocked }) => (
  <FieldSection title="Advanced: Match Rules">
    <UrlPatternField
      name="extensionPoint.definition.isAvailable.urlPatterns"
      {...makeLockableFieldProps("URL Patterns", isLocked)}
    />

    <SelectorMatchField
      name="extensionPoint.definition.isAvailable.selectors"
      {...makeLockableFieldProps("Selectors", isLocked)}
    />
  </FieldSection>
);

export default MatchRulesSection;
