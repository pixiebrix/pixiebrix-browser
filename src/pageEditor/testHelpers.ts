/*
 * Copyright (C) 2023 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import { configureStore } from "@reduxjs/toolkit";
import { authSlice } from "@/auth/authSlice";
import extensionsSlice from "@/store/extensionsSlice";
import servicesSlice from "@/store/servicesSlice";
import settingsSlice from "@/store/settingsSlice";
import { editorSlice } from "@/pageEditor/slices/editorSlice";
import sessionSlice from "@/pageEditor/slices/sessionSlice";
import { savingExtensionSlice } from "@/pageEditor/panes/save/savingExtensionSlice";
import runtimeSlice from "@/pageEditor/slices/runtimeSlice";
import { logSlice } from "@/components/logViewer/logSlice";
import {
  createRenderHookWithWrappers,
  createRenderWithWrappers,
} from "@/testUtils/testHelpers";
import analysisSlice from "@/analysis/analysisSlice";
import pageEditorAnalysisManager from "./analysisManager";
import { tabStateSlice } from "@/pageEditor/tabState/tabStateSlice";
import { appApi } from "@/services/api";
import { modDefinitionsSlice } from "@/modDefinitions/modDefinitionsSlice";
import { modDefinitionsMiddleware } from "@/modDefinitions/modDefinitionsListenerMiddleware";

const configureStoreForTests = () =>
  configureStore({
    reducer: {
      auth: authSlice.reducer,
      options: extensionsSlice.reducer,
      services: servicesSlice.reducer,
      settings: settingsSlice.reducer,
      editor: editorSlice.reducer,
      session: sessionSlice.reducer,
      savingExtension: savingExtensionSlice.reducer,
      runtime: runtimeSlice.reducer,
      logs: logSlice.reducer,
      analysis: analysisSlice.reducer,
      tabState: tabStateSlice.reducer,
      modDefinitions: modDefinitionsSlice.reducer,
      [appApi.reducerPath]: appApi.reducer,
    },
    middleware(getDefaultMiddleware) {
      /* eslint-disable unicorn/prefer-spread -- It's not Array#concat, can't use spread */
      return getDefaultMiddleware()
        .concat(appApi.middleware)
        .concat(pageEditorAnalysisManager.middleware)
        .concat(modDefinitionsMiddleware);
      /* eslint-enable unicorn/prefer-spread */
    },
  });

const renderWithWrappers = createRenderWithWrappers(configureStoreForTests);
const renderHookWithWrappers = createRenderHookWithWrappers(
  configureStoreForTests
);

// eslint-disable-next-line import/export -- re-export RTL
export * from "@testing-library/react";
// eslint-disable-next-line import/export -- override render
export { renderWithWrappers as render };
export { renderHookWithWrappers as renderHook };
