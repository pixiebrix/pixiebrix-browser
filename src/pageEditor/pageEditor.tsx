/*
 * Copyright (C) 2024 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import "@/vendors/bootstrapWithoutRem.css";
import "@/utils/global.scss";
import "@/utils/layout.scss";

import "@/extensionContext";
import "@/development/darkMode.js";

import { initMessengerLogging } from "@/development/messengerLogging";

import ReactDOM from "react-dom";
import React from "react";
import Panel from "@/pageEditor/Panel";
import { watchNavigation } from "@/pageEditor/protocol";
import { initToaster } from "@/utils/notify";
import { initRuntimeLogging } from "@/development/runtimeLogging";
import { initPerformanceMonitoring } from "@/telemetry/performance";
import { setPlatform } from "@/platform/platformContext";
import extensionPagePlatform from "@/extensionPages/extensionPagePlatform";

async function init() {
  setPlatform(extensionPagePlatform);
  void initMessengerLogging();
  void initRuntimeLogging();
  try {
    await initPerformanceMonitoring();
  } catch (error) {
    console.error("Failed to initialize performance monitoring", error);
  }

  watchNavigation();
  initToaster();

  ReactDOM.render(<Panel />, document.querySelector("#container"));
}

void init();
