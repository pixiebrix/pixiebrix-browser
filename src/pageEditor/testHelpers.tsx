/*
 * Copyright (C) 2022 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import { configureStore } from "@reduxjs/toolkit";
import { authSlice } from "@/auth/authSlice";
import extensionsSlice from "@/store/extensionsSlice";
import servicesSlice from "@/store/servicesSlice";
import settingsSlice from "@/store/settingsSlice";
import { editorSlice } from "@/pageEditor/slices/editorSlice";
import sessionSlice from "@/pageEditor/slices/sessionSlice";
import { savingExtensionSlice } from "@/pageEditor/panes/save/savingExtensionSlice";
import runtimeSlice from "@/pageEditor/slices/runtimeSlice";
import { logSlice } from "@/components/logViewer/logSlice";
import { createRenderWithWrappers } from "@/testUtils/testHelpers";
import { appApi } from "@/services/api";

const renderWithWrappers = createRenderWithWrappers(() =>
  configureStore({
    reducer: {
      auth: authSlice.reducer,
      options: extensionsSlice.reducer,
      services: servicesSlice.reducer,
      settings: settingsSlice.reducer,
      editor: editorSlice.reducer,
      session: sessionSlice.reducer,
      savingExtension: savingExtensionSlice.reducer,
      runtime: runtimeSlice.reducer,
      logs: logSlice.reducer,
      [appApi.reducerPath]: appApi.reducer,
    },
  })
);

// eslint-disable-next-line import/export -- re-export RTL
export * from "@testing-library/react";
// eslint-disable-next-line import/export -- override render
export { renderWithWrappers as render };
