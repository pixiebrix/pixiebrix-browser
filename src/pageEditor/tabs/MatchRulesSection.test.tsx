/*
 * Copyright (C) 2023 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import React from "react";
import { render } from "@/pageEditor/testHelpers";
import registerDefaultWidgets from "@/components/fields/schemaFields/widgets/registerDefaultWidgets";
import MatchRulesSection from "@/pageEditor/tabs/MatchRulesSection";
import { set } from "lodash";

beforeEach(() => {
  registerDefaultWidgets();
});

describe("MatchRulesSection", () => {
  test("render empty section", () => {
    expect(
      render(<MatchRulesSection isLocked={false} />, {
        initialValues: set({}, "extensionPoint.definition.isAvailable", {
          matchPatterns: [],
          urlPatterns: [],
          selectors: [],
        }),
      }).asFragment()
    ).toMatchSnapshot();
  });

  test("render populated section", () => {
    expect(
      render(<MatchRulesSection isLocked={false} />, {
        initialValues: set({}, "extensionPoint.definition.isAvailable", {
          matchPatterns: [],
          urlPatterns: [
            {
              hash: "/foo/",
            },
          ],
          selectors: ["#foo"],
        }),
        setupRedux(dispatch) {},
      }).asFragment()
    ).toMatchSnapshot();
  });
});
