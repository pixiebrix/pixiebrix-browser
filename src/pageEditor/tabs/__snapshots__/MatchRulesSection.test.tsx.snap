// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MatchRulesSection render empty section 1`] = `
<DocumentFragment>
  <form
    action="#"
  >
    <div
      class="root"
    >
      <button
        class="header"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-chevron-right fa-w-10 activeIndicator active"
          data-icon="chevron-right"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 320 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M285.476 272.971L91.132 467.314c-9.373 9.373-24.569 9.373-33.941 0l-22.667-22.667c-9.357-9.357-9.375-24.522-.04-33.901L188.505 256 34.484 101.255c-9.335-9.379-9.317-24.544.04-33.901l22.667-22.667c9.373-9.373 24.569-9.373 33.941 0L285.475 239.03c9.373 9.372 9.373 24.568.001 33.941z"
            fill="currentColor"
          />
        </svg>
        Advanced: Match Rules
      </button>
      <div
        class="body collapse show"
      >
        <div
          class="formGroup form-group row"
        >
          <label
            class="label form-label col-form-label col-xl-2 col-lg-3"
            for="extensionPoint.definition.isAvailable.urlPatterns"
          >
            URL Patterns
          </label>
          <div
            class="col-xl-10 col-lg-9"
          >
            <ul
              class="list-group mb-2"
            />
            <button
              class="addButton btn btn-link"
              type="button"
            >
              Add URL Pattern
            </button>
            <small
              class="text-muted form-text"
            >
              <span>
                URL pattern rules restricting when this starter brick runs. If provided, at least one of the rules must match for this starter brick to run. See 
                <a
                  href="https://developer.mozilla.org/en-US/docs/Web/API/URL_Pattern_API"
                  rel="noreferrer"
                  target="_blank"
                >
                  URL Pattern API Documentation
                </a>
                 for examples.
              </span>
            </small>
          </div>
        </div>
        <div
          class="formGroup form-group row"
        >
          <label
            class="label form-label col-form-label col-xl-2 col-lg-3"
            for="extensionPoint.definition.isAvailable.selectors"
          >
            Selectors
          </label>
          <div
            class="col-xl-10 col-lg-9"
          >
            <ul
              class="list-group mb-2"
            />
            <button
              class="addButton btn btn-link"
              type="button"
            >
              Add Selector
            </button>
            <small
              class="text-muted form-text"
            >
              <span>
                Selectors restricting when this starter brick runs. If provided, at least one of the selectors must match 
                <i>
                  on page load
                </i>
                 for this starter brick to run.
              </span>
            </small>
          </div>
        </div>
      </div>
    </div>
    <button
      type="submit"
    >
      Submit
    </button>
  </form>
</DocumentFragment>
`;

exports[`MatchRulesSection render populated section 1`] = `
<DocumentFragment>
  <form
    action="#"
  >
    <div
      class="root"
    >
      <button
        class="header"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-chevron-right fa-w-10 activeIndicator active"
          data-icon="chevron-right"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 320 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M285.476 272.971L91.132 467.314c-9.373 9.373-24.569 9.373-33.941 0l-22.667-22.667c-9.357-9.357-9.375-24.522-.04-33.901L188.505 256 34.484 101.255c-9.335-9.379-9.317-24.544.04-33.901l22.667-22.667c9.373-9.373 24.569-9.373 33.941 0L285.475 239.03c9.373 9.372 9.373 24.568.001 33.941z"
            fill="currentColor"
          />
        </svg>
        Advanced: Match Rules
      </button>
      <div
        class="body collapse show"
      >
        <div
          class="formGroup form-group row"
        >
          <label
            class="label form-label col-form-label col-xl-2 col-lg-3"
            for="extensionPoint.definition.isAvailable.urlPatterns"
          >
            URL Patterns
          </label>
          <div
            class="col-xl-10 col-lg-9"
          >
            <ul
              class="list-group mb-2"
            >
              <li
                class="list-group-item py-1"
              >
                <div
                  class="formGroup mb-0 form-group row"
                >
                  <div
                    class="col-xl-12 col-lg-12"
                  >
                    <div
                      class="root"
                    >
                      <div
                        class="field"
                      >
                        <div>
                          <table
                            class="mb-0 table table-sm"
                          >
                            <thead>
                              <tr>
                                <th
                                  scope="col"
                                >
                                  Property name
                                </th>
                                <th
                                  scope="col"
                                >
                                  Value
                                </th>
                              </tr>
                            </thead>
                            <tbody>
                              <tr>
                                <td>
                                  <input
                                    class="form-control"
                                    data-testid="extensionPoint.definition.isAvailable.urlPatterns.0.hostname-name"
                                    readonly=""
                                    type="text"
                                    value="hostname"
                                  />
                                </td>
                                <td>
                                  <div
                                    class="formGroup mb-0 form-group row"
                                  >
                                    <div
                                      class="col-xl-12 col-lg-12"
                                    >
                                      <div
                                        class="root"
                                      >
                                        <div
                                          class="field"
                                        >
                                          <input
                                            class="form-control"
                                            id="extensionPoint.definition.isAvailable.urlPatterns.0.hostname"
                                            name="extensionPoint.definition.isAvailable.urlPatterns.0.hostname"
                                            readonly=""
                                            value=""
                                          />
                                        </div>
                                        <div
                                          class="dropdown dropdown"
                                          data-test-selected="Exclude"
                                          data-testid="toggle-extensionPoint.definition.isAvailable.urlPatterns.0.hostname"
                                        >
                                          <button
                                            aria-expanded="false"
                                            aria-haspopup="true"
                                            class="dropdown-toggle btn btn-link"
                                            type="button"
                                          >
                                            <span
                                              class="symbol"
                                            >
                                              <test-file-stub
                                                classname="root"
                                              />
                                            </span>
                                          </button>
                                        </div>
                                      </div>
                                      <small
                                        class="text-muted form-text"
                                      >
                                        <span>
                                          Pattern to match the hostname part of a URL.
                                        </span>
                                      </small>
                                    </div>
                                  </div>
                                </td>
                              </tr>
                              <tr>
                                <td>
                                  <input
                                    class="form-control"
                                    data-testid="extensionPoint.definition.isAvailable.urlPatterns.0.pathname-name"
                                    readonly=""
                                    type="text"
                                    value="pathname"
                                  />
                                </td>
                                <td>
                                  <div
                                    class="formGroup mb-0 form-group row"
                                  >
                                    <div
                                      class="col-xl-12 col-lg-12"
                                    >
                                      <div
                                        class="root"
                                      >
                                        <div
                                          class="field"
                                        >
                                          <input
                                            class="form-control"
                                            id="extensionPoint.definition.isAvailable.urlPatterns.0.pathname"
                                            name="extensionPoint.definition.isAvailable.urlPatterns.0.pathname"
                                            readonly=""
                                            value=""
                                          />
                                        </div>
                                        <div
                                          class="dropdown dropdown"
                                          data-test-selected="Exclude"
                                          data-testid="toggle-extensionPoint.definition.isAvailable.urlPatterns.0.pathname"
                                        >
                                          <button
                                            aria-expanded="false"
                                            aria-haspopup="true"
                                            class="dropdown-toggle btn btn-link"
                                            type="button"
                                          >
                                            <span
                                              class="symbol"
                                            >
                                              <test-file-stub
                                                classname="root"
                                              />
                                            </span>
                                          </button>
                                        </div>
                                      </div>
                                      <small
                                        class="text-muted form-text"
                                      >
                                        <span>
                                          Pattern to match the pathname part of a URL.
                                        </span>
                                      </small>
                                    </div>
                                  </div>
                                </td>
                              </tr>
                              <tr>
                                <td>
                                  <input
                                    class="form-control"
                                    data-testid="extensionPoint.definition.isAvailable.urlPatterns.0.hash-name"
                                    readonly=""
                                    type="text"
                                    value="hash"
                                  />
                                </td>
                                <td>
                                  <div
                                    class="formGroup mb-0 form-group row"
                                  >
                                    <div
                                      class="col-xl-12 col-lg-12"
                                    >
                                      <div
                                        class="root"
                                      >
                                        <div
                                          class="field"
                                        >
                                          <textarea
                                            class="form-control"
                                            id="extensionPoint.definition.isAvailable.urlPatterns.0.hash"
                                            name="extensionPoint.definition.isAvailable.urlPatterns.0.hash"
                                            rows="1"
                                          >
                                            /foo/
                                          </textarea>
                                        </div>
                                        <div
                                          class="dropdown dropdown"
                                          data-test-selected="Text"
                                          data-testid="toggle-extensionPoint.definition.isAvailable.urlPatterns.0.hash"
                                        >
                                          <button
                                            aria-expanded="false"
                                            aria-haspopup="true"
                                            class="dropdown-toggle btn btn-link"
                                            type="button"
                                          >
                                            <span
                                              class="symbol"
                                            >
                                              <test-file-stub
                                                classname="root"
                                              />
                                            </span>
                                          </button>
                                        </div>
                                      </div>
                                      <small
                                        class="text-muted form-text"
                                      >
                                        <span>
                                          Pattern to match the hash part of a URL.
                                        </span>
                                      </small>
                                    </div>
                                  </div>
                                </td>
                              </tr>
                              <tr>
                                <td>
                                  <input
                                    class="form-control"
                                    data-testid="extensionPoint.definition.isAvailable.urlPatterns.0.search-name"
                                    readonly=""
                                    type="text"
                                    value="search"
                                  />
                                </td>
                                <td>
                                  <div
                                    class="formGroup mb-0 form-group row"
                                  >
                                    <div
                                      class="col-xl-12 col-lg-12"
                                    >
                                      <div
                                        class="root"
                                      >
                                        <div
                                          class="field"
                                        >
                                          <input
                                            class="form-control"
                                            id="extensionPoint.definition.isAvailable.urlPatterns.0.search"
                                            name="extensionPoint.definition.isAvailable.urlPatterns.0.search"
                                            readonly=""
                                            value=""
                                          />
                                        </div>
                                        <div
                                          class="dropdown dropdown"
                                          data-test-selected="Exclude"
                                          data-testid="toggle-extensionPoint.definition.isAvailable.urlPatterns.0.search"
                                        >
                                          <button
                                            aria-expanded="false"
                                            aria-haspopup="true"
                                            class="dropdown-toggle btn btn-link"
                                            type="button"
                                          >
                                            <span
                                              class="symbol"
                                            >
                                              <test-file-stub
                                                classname="root"
                                              />
                                            </span>
                                          </button>
                                        </div>
                                      </div>
                                      <small
                                        class="text-muted form-text"
                                      >
                                        <span>
                                          Pattern to match the search part of a URL.
                                        </span>
                                      </small>
                                    </div>
                                  </div>
                                </td>
                              </tr>
                            </tbody>
                          </table>
                          <button
                            class="btn btn-primary"
                            type="button"
                          >
                            Add Property
                          </button>
                        </div>
                      </div>
                      <div
                        class="dropdown dropdown"
                        data-test-selected="Object properties"
                        data-testid="toggle-extensionPoint.definition.isAvailable.urlPatterns.0"
                      >
                        <button
                          aria-expanded="false"
                          aria-haspopup="true"
                          class="dropdown-toggle btn btn-link"
                          type="button"
                        >
                          <span
                            class="symbol"
                          >
                            <test-file-stub
                              classname="root"
                            />
                          </span>
                        </button>
                      </div>
                    </div>
                  </div>
                </div>
              </li>
            </ul>
            <button
              class="addButton btn btn-link"
              type="button"
            >
              Add URL Pattern
            </button>
            <small
              class="text-muted form-text"
            >
              <span>
                URL pattern rules restricting when this starter brick runs. If provided, at least one of the rules must match for this starter brick to run. See 
                <a
                  href="https://developer.mozilla.org/en-US/docs/Web/API/URL_Pattern_API"
                  rel="noreferrer"
                  target="_blank"
                >
                  URL Pattern API Documentation
                </a>
                 for examples.
              </span>
            </small>
          </div>
        </div>
        <div
          class="formGroup form-group row"
        >
          <label
            class="label form-label col-form-label col-xl-2 col-lg-3"
            for="extensionPoint.definition.isAvailable.selectors"
          >
            Selectors
          </label>
          <div
            class="col-xl-10 col-lg-9"
          >
            <ul
              class="list-group mb-2"
            >
              <li
                class="list-group-item py-1"
              >
                <div
                  class="formGroup mb-0 form-group row"
                >
                  <div
                    class="col-xl-12 col-lg-12"
                  >
                    <div
                      class="root"
                    >
                      <div
                        class="field"
                      >
                        <textarea
                          class="form-control"
                          id="extensionPoint.definition.isAvailable.selectors.0"
                          name="extensionPoint.definition.isAvailable.selectors.0"
                          rows="1"
                        >
                          #foo
                        </textarea>
                      </div>
                      <div
                        class="dropdown dropdown"
                        data-test-selected="Text"
                        data-testid="toggle-extensionPoint.definition.isAvailable.selectors.0"
                      >
                        <button
                          aria-expanded="false"
                          aria-haspopup="true"
                          class="dropdown-toggle btn btn-link"
                          type="button"
                        >
                          <span
                            class="symbol"
                          >
                            <test-file-stub
                              classname="root"
                            />
                          </span>
                        </button>
                      </div>
                    </div>
                  </div>
                </div>
              </li>
            </ul>
            <button
              class="addButton btn btn-link"
              type="button"
            >
              Add Selector
            </button>
            <small
              class="text-muted form-text"
            >
              <span>
                Selectors restricting when this starter brick runs. If provided, at least one of the selectors must match 
                <i>
                  on page load
                </i>
                 for this starter brick to run.
              </span>
            </small>
          </div>
        </div>
      </div>
    </div>
    <button
      type="submit"
    >
      Submit
    </button>
  </form>
</DocumentFragment>
`;
